cpp/src/arrow/python/deserialize.cc

```diff
@@ -195,8 +196,10 @@ std::vector<int8_t> GetPythonTypes(const UnionArray& data) {
   std::vector<int8_t> result;
   auto type = data.type();
   for (int i = 0; i < type->num_children(); ++i) {
-    // stoi is locale dependent, but should be ok for small integers
-    result.push_back(static_cast<int8_t>(std::stoi(type->child(i)->name())));
+    std::istringstream convert(type->child(i)->name());
+    int tag;
+    convert >> tag;
```

Comment:
This will have the same problem as `std::stoi`. Have you looked at `utils/parsing.h`? The various StringConverter classes there could help ;-)